facebook.com: Unable to drag and drop images (from Photos, Finder) into new album

whsieh opened the pull request:
#### ddcd91a27bd801bebc876705a4d6e1e062f9ad2e
<pre>
facebook.com: Unable to drag and drop images (from Photos, Finder) into new album
<a href="https://bugs.webkit.org/show_bug.cgi?id=301472">https://bugs.webkit.org/show_bug.cgi?id=301472</a>
<a href="https://rdar.apple.com/150848219">rdar://150848219</a>

Reviewed by Abrar Rahman Protyasha.

When dragging an image from either Photos or Finder and dropping into the album creation UI on
facebook.com in Safari, the file drop is effectively ignored. This happens because Facebook&apos;s script
expects the `dragenter` event to contain a non-empty list of DataTransfer items, with:

1. `item.kind` equal to `&quot;file&quot;`
2. A non-empty `item.type` that&apos;s not `text/html`, `text/plain`, or `text/uri-list`.

WebKit currently exposes an empty list of items (due to restrictions on reading drag items before
the drop is being performed), which causes Facebook&apos;s script to avoid calling `preventDefault()` on
`dragenter`. Both Blink and Gecko expose a list of data transfer items in this case that contains
the dropped file(s) and their MIME types, but attempts to get the file (`getAsFile()`) before the
`drop` event has been fired will result in a null `File` object. This patch aligns WebKit&apos;s behavior
to the other engines (with some caveats — see below), which fixes this bug on Facebook.

Tests: added to DragAndDropTests.DragPromisedImageFileIntoFileUpload

Tests: Tools/TestWebKitAPI/Tests/WebKitCocoa/image-and-file-upload.html
       Tools/TestWebKitAPI/Tests/mac/DragAndDropTestsMac.mm
       Tools/TestWebKitAPI/mac/TestDraggingInfo.mm
* Source/WebCore/dom/DataTransferItem.cpp:
(WebCore::DataTransferItem::create):

Add a `Kind` argument to this creation method, to create a `DataTransferItem` with a given type,
representing either a `String` or a placeholder for a `File`. Use this when dispatching drag update
events to expose file data transfer items that only expose type information.

(WebCore::DataTransferItem::DataTransferItem):
(WebCore::DataTransferItem::isFile const):
(WebCore::DataTransferItem::kind const):
* Source/WebCore/dom/DataTransferItem.h:

Introduce an `m_kind` member variable, and use it for the `kind()` attribute instead of basing it
off of the presence of `m_file`. This is because `file` items may now have a null `m_file`, before
the `drop` event is dispatched.

(WebCore::DataTransferItem::isFile const): Deleted.
* Source/WebCore/dom/DataTransferItemList.cpp:
(WebCore::DataTransferItemList::add):
(WebCore::DataTransferItemList::ensureItems const):

Apply the main fix here — only expose `File`-backed `DataTransferItem`s when the `DataTransfer` can
read data; when the `DataTransfer` can only read types, we instead expose a list of items based on
the promised MIME types.

In order to reduce fingerprinting surface when the user drags over unintended drop targets, we
additionally only expose known image and media types.

(WebCore::DataTransferItemList::didSetStringData):
* Source/WebCore/platform/DragData.h:
(WebCore::DragData::DragData):
(WebCore::DragData::setPromisedFileMIMETypes):
(WebCore::DragData::promisedFileMIMETypes const):

Add a list of MIME types of promised files when dragging over drop targets.

* Source/WebCore/platform/Pasteboard.h:
(WebCore::Pasteboard::Pasteboard):
(WebCore::Pasteboard::promisedFileMIMETypes const):
* Source/WebCore/platform/cocoa/DragDataCocoa.mm:
(WebCore::DragData::DragData):
* Source/WebCore/platform/mac/PasteboardMac.mm:
(WebCore::Pasteboard::Pasteboard):
(WebCore::Pasteboard::create):

Plumb the list of promised file MIME types from `DragData` → `Pasteboard`.

* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/UIProcess/API/mac/WKWebViewMac.h:
* Source/WebKit/UIProcess/API/mac/WKWebViewMac.mm:
(-[WKWebView _promisedFileMIMETypes:]):

Add a helper method to extract a list of MIME types for promised files in the drag session, before
the drag operation (drop) is performed. We get this via `-[NSFilePromiseReceiver fileTypes]` on the
dragged items, and otherwise fall back to getting preferred MIME types from dragged file URLs.

* Source/WebKit/UIProcess/mac/WebViewImpl.mm:
(WebKit::applicationFlagsForDrag):
(WebKit::WebViewImpl::draggingEntered):
(WebKit::WebViewImpl::draggingUpdated):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/image-and-file-upload.html:
* Tools/TestWebKitAPI/Tests/mac/DragAndDropTestsMac.mm:
(TEST(DragAndDropTests, DragPromisedImageFileIntoFileUpload)):

Augment an existing API test to also exercise this scenario by dumping the contents of the
`e.dataTransfer.items` while dragging.

* Tools/TestWebKitAPI/mac/TestDraggingInfo.mm:
(-[TestDraggingInfo enumerateDraggingItemsWithOptions:forView:classes:searchOptions:usingBlock:]):

Adjust this testing helper class to lazily generate `NSFilePromiseReceiver`s when WebKit asks to
`-enumerateDraggingItemsWithOptions:forView:classes:searchOptions:usingBlock:` to ensure that we
don&apos;t end up with dozens of redundant file promise receivers when simulating a drop. This was
previously not necessary because we only called this once when performing a drop, but we now need to
call it continually while dragging over the view.

Canonical link: <a href="https://commits.webkit.org/302156@main">https://commits.webkit.org/302156@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/251400c262459e84fa8aba5cc1ebcbb55b409707

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/128174 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/457 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/39006 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/135567 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/79662 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/8a1c0c92-8bc1-4975-a9f8-741296bd5226) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/130046 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/382 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/329 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/97568 "Passed tests") | [❌ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/65466 "") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/b84f3cad-b2c0-4440-826e-65dd68b08e45) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/131122 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/237 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/114826 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/78139 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/d50aa851-6cb0-4b73-bdbd-a0878b2d12a6) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/227 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/32934 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/78846 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/108604 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/33418 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/138025 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/309 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/294 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/106097 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/341 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/111168 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/105881 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/26984 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/242 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/29722 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/52553 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/355 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/62444 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/262 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/336 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/320 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

aprotyas commented on Source/WebCore/dom/DataTransferItem.h:
```diff
@@ -49,29 +49,31 @@ class StringCallback;
 
     WeakPtr<DataTransferItemList> m_list;
     const String m_type;
+    const String m_kind;
```
Can kind be anything other than `string`/`file`? I'm wondering if you can back this as an enum and convert to the string representation when needed (in `DataTransferItem::kind()`)

aprotyas replied:
Something like `enum class DataTransferItem::Kind : uint8_t { String, File }` and you can plumb that through with `static Ref<DataTransferItem> create(WeakPtr<DataTransferItemList>&&, const String&, Kind);`

whsieh replied:
Makes sense — it should only be `string` / `file`. Thanks for the review!

whsieh committed:
facebook.com: Unable to drag and drop images (from Photos, Finder) into new album
https://bugs.webkit.org/show_bug.cgi?id=301472
rdar://150848219

Reviewed by Abrar Rahman Protyasha.

When dragging an image from either Photos or Finder and dropping into the album creation UI on
facebook.com in Safari, the file drop is effectively ignored. This happens because Facebook's script
expects the `dragenter` event to contain a non-empty list of DataTransfer items, with:

1. `item.kind` equal to `"file"`
2. A non-empty `item.type` that's not `text/html`, `text/plain`, or `text/uri-list`.

WebKit currently exposes an empty list of items (due to restrictions on reading drag items before
the drop is being performed), which causes Facebook's script to avoid calling `preventDefault()` on
`dragenter`. Both Blink and Gecko expose a list of data transfer items in this case that contains
the dropped file(s) and their MIME types, but attempts to get the file (`getAsFile()`) before the
`drop` event has been fired will result in a null `File` object. This patch aligns WebKit's behavior
to the other engines (with some caveats — see below), which fixes this bug on Facebook.

Tests: added to DragAndDropTests.DragPromisedImageFileIntoFileUpload

Tests: Tools/TestWebKitAPI/Tests/WebKitCocoa/image-and-file-upload.html
       Tools/TestWebKitAPI/Tests/mac/DragAndDropTestsMac.mm
       Tools/TestWebKitAPI/mac/TestDraggingInfo.mm
* Source/WebCore/dom/DataTransferItem.cpp:
(WebCore::DataTransferItem::create):

Add a `Kind` argument to this creation method, to create a `DataTransferItem` with a given type,
representing either a `String` or a placeholder for a `File`. Use this when dispatching drag update
events to expose file data transfer items that only expose type information.

(WebCore::DataTransferItem::DataTransferItem):
(WebCore::DataTransferItem::isFile const):
(WebCore::DataTransferItem::kind const):
* Source/WebCore/dom/DataTransferItem.h:

Introduce an `m_kind` member variable, and use it for the `kind()` attribute instead of basing it
off of the presence of `m_file`. This is because `file` items may now have a null `m_file`, before
the `drop` event is dispatched.

(WebCore::DataTransferItem::isFile const): Deleted.
* Source/WebCore/dom/DataTransferItemList.cpp:
(WebCore::DataTransferItemList::add):
(WebCore::DataTransferItemList::ensureItems const):

Apply the main fix here — only expose `File`-backed `DataTransferItem`s when the `DataTransfer` can
read data; when the `DataTransfer` can only read types, we instead expose a list of items based on
the promised MIME types.

In order to reduce fingerprinting surface when the user drags over unintended drop targets, we
additionally only expose known image and media types.

(WebCore::DataTransferItemList::didSetStringData):
* Source/WebCore/platform/DragData.h:
(WebCore::DragData::DragData):
(WebCore::DragData::setPromisedFileMIMETypes):
(WebCore::DragData::promisedFileMIMETypes const):

Add a list of MIME types of promised files when dragging over drop targets.

* Source/WebCore/platform/Pasteboard.h:
(WebCore::Pasteboard::Pasteboard):
(WebCore::Pasteboard::promisedFileMIMETypes const):
* Source/WebCore/platform/cocoa/DragDataCocoa.mm:
(WebCore::DragData::DragData):
* Source/WebCore/platform/mac/PasteboardMac.mm:
(WebCore::Pasteboard::Pasteboard):
(WebCore::Pasteboard::create):

Plumb the list of promised file MIME types from `DragData` → `Pasteboard`.

* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/UIProcess/API/mac/WKWebViewMac.h:
* Source/WebKit/UIProcess/API/mac/WKWebViewMac.mm:
(-[WKWebView _promisedFileMIMETypes:]):

Add a helper method to extract a list of MIME types for promised files in the drag session, before
the drag operation (drop) is performed. We get this via `-[NSFilePromiseReceiver fileTypes]` on the
dragged items, and otherwise fall back to getting preferred MIME types from dragged file URLs.

* Source/WebKit/UIProcess/mac/WebViewImpl.mm:
(WebKit::applicationFlagsForDrag):
(WebKit::WebViewImpl::draggingEntered):
(WebKit::WebViewImpl::draggingUpdated):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/image-and-file-upload.html:
* Tools/TestWebKitAPI/Tests/mac/DragAndDropTestsMac.mm:
(TEST(DragAndDropTests, DragPromisedImageFileIntoFileUpload)):

Augment an existing API test to also exercise this scenario by dumping the contents of the
`e.dataTransfer.items` while dragging.

* Tools/TestWebKitAPI/mac/TestDraggingInfo.mm:
(-[TestDraggingInfo enumerateDraggingItemsWithOptions:forView:classes:searchOptions:usingBlock:]):

Adjust this testing helper class to lazily generate `NSFilePromiseReceiver`s when WebKit asks to
`-enumerateDraggingItemsWithOptions:forView:classes:searchOptions:usingBlock:` to ensure that we
don't end up with dozens of redundant file promise receivers when simulating a drop. This was
previously not necessary because we only called this once when performing a drop, but we now need to
call it continually while dragging over the view.

Canonical link: https://commits.webkit.org/302156@main
